src/backend/distributed/utils/citus_clauses.c

```diff
@@ -45,6 +45,11 @@ static bool ShouldEvaluateFunctionWithMasterContext(MasterEvaluationContext *
 bool
 RequiresMasterEvaluation(Query *query)
 {
+	if (query->commandType == CMD_SELECT)
```

Comment:
There's a general question of how function evaluation should work in modifying CTEs. We certainly cannot run it on the outer SELECT query. 

The contain_dml walker seems somewhat redundant, since we can just check `hasModifyingCTE` if we want to know whether there are modifying CTEs. They can only appear at the top level.

Maybe `if (query->commandType == CMD_SELECT && !query->hasModifyingCTE)`? I would feel it is somewhat confusing since the latter suggests functionality that is not there.